### ISSUE TITLE
LE: Rate limiting woes: some certificates not saved

### ISSUE THREAD

**Issue by kseistrup** [mholt/caddy#383]
I run a multi-host setup but I didn't know about Let's Encrypt's [public beta rate limits](https://community.letsencrypt.org/t/public-beta-rate-limits/4772), so the first time I ran Caddy 0.8 it failed for some of the domains that have several subdomains:
```txt
[subdom.example.com] failed to get certificate: acme: Error 429 - urn:acme:error:rateLimited - Error creating new cert :: Too many certificates already issued for: example.com
```
This happened even if I pruned the list to be within the acceptable limits. However, I still receive the same error messages, and it seems it's because Caddy saves one certificate for each domain name it sees, but only saves these if there were no errors at all. So the first time I ran Caddy it apparrently registered as many domains as it could, but didn't save any of these due to some errors from Let's Encrypt. Now Let's Encrypt refuses to create more subdomains because it thinks it has already created them, but Caddy hasn't saved them anywhere.

Caddy managed to save two certificates from a third run, and it seems indeed that it requests one certificate for each domain it sees.

**Comment by lenovouser** [comment]
The rate limit error showed up to me on my first try using Let's Encrypt too. It has nothing to do with Caddy. As a temporary solution, disable IPv6 with:

```
sudo sysctl -w net.ipv6.conf.all.disable_ipv6=1
```

issue your Let's Encrypt certificate,

and re-enable IPv6 with:
```
sudo sysctl -w net.ipv6.conf.all.disable_ipv6=0
```

The above works on Ubuntu/Debian.

**Comment by kseistrup** [comment]
How come IPv6 is a factor in this context?

**Comment by andmarios** [comment]
I hit the same issue. Due to some tests I restarted caddy a few times. Each time it requested (and used) a new certificate instead of saving the old one, thus I got my domain rate-limited.

I was on the previous commit (fd14f257) than the current one (5b93799) but the diff doesn't show any important differences that would affect the result.

Disclaimer: I am new to caddy, so there may be some negligence on my part.

**Comment by lenovouser** [comment]
As I just said, this is probably not Caddy's fault. More information on IPv6 here: [community.letsencrypt.org/unexpected-registration-rate-limiting-error](https://community.letsencrypt.org/t/unexpected-registration-rate-limiting-error/2157)

**Comment by kseistrup** [comment]
Thanks :+1:

**Comment by kseistrup** [comment]
Except the error I get is not “Too many registrations from this IP”, but “Too many certificates already issued for […]”. So perhaps we're talking about two issues here?

**Comment by andmarios** [comment]
And as we just said this isn't a rate-limit per IP but per domain.

Acmetool is able to get me certificates fine except for the domain that got blacklisted due to caddy.

**Comment by mholt** [comment]
We could probably handle the rate limiting better; I didn't expect it to be so low at first, and especially for the rate limit to cover all subdomains too.

As has been mentioned, the IPv6 issue is a limitation of Let's Encrypt's infrastructure. Hopefully this is fixed soon, but that's all we can do at this point.

Once rate limited, you'll have to wait 7 days for that domain and any of its subdomains. Again, hopefully this is lifted in the near future!

I'll work on including some better error handling in the next patch release so that it at least stores what certificates it has to disk.

**Comment by Moter8** [comment]
Hit this too today:

```
ubuntu@ip-172-31-21-209:~/caddy$ caddy -log stderr
Activating privacy features...2015/12/05 21:48:08 [INFO] acme: Obtaining bundled certificates for bwf.ninja, buildapc.xyz, moter8.xyz, chag.xyz, chagemann.de, rainbow.moter8.xyz
2015/12/05 21:48:09 [INFO] acme: Trying to solve TLS-SNI-01
2015/12/05 21:48:11 The server validated our request
2015/12/05 21:48:11 [INFO] acme: Trying to solve TLS-SNI-01
2015/12/05 21:48:14 The server validated our request
2015/12/05 21:48:14 [INFO] acme: Trying to solve HTTP-01
2015/12/05 21:48:20 [INFO] acme: Trying to solve TLS-SNI-01
2015/12/05 21:48:22 The server validated our request
2015/12/05 21:48:22 [INFO] acme: Trying to solve HTTP-01
2015/12/05 21:48:28 [INFO] acme: Trying to solve TLS-SNI-01
2015/12/05 21:48:31 The server validated our request
2015/12/05 21:48:31 [INFO] acme: Validations succeeded; requesting certificates
2015/12/05 21:48:32 [INFO] acme: Requesting issuer cert from https://acme-v01.api.letsencrypt.org/acme/issuer-cert
2015/12/05 21:48:32 [buildapc.xyz] Server responded with a certificate.
2015/12/05 21:48:32 [INFO] acme: Requesting issuer cert from https://acme-v01.api.letsencrypt.org/acme/issuer-cert
2015/12/05 21:48:32 [bwf.ninja] Server responded with a certificate.
2015/12/05 21:48:32 [INFO] acme: Requesting issuer cert from https://acme-v01.api.letsencrypt.org/acme/issuer-cert
2015/12/05 21:48:32 [chag.xyz] Server responded with a certificate.

2015/12/05 21:48:32 [chagemann.de] failed to get certificate: The server could not validate our request.
[rainbow.moter8.xyz] failed to get certificate: acme: Error 429 - urn:acme:error:rateLimited - Error creating new cert :: Too many certificates already issued for: moter8.xyz
[moter8.xyz] failed to get certificate: The server could not validate our request.
```

Sometimes it just said "The server could not validate our request" and after executing the command again it then informed me about the rate limit. All domains are tempbanned now :D

**Comment by kseistrup** [comment]
How do we stop Caddy from requesting certificates that are already present on disk?  I had to restart Caddy due to a recompile (latest git), but then it started requesting certificates it had already saved some days ago.

**Comment by mholt** [comment]
Caddy doesn't request certificates it already has. Are you running it as the same user? Is it for the exact same hostnames? Did the certificates still exist on disk?

**Comment by kseistrup** [comment]
I'm absolute sure that the certificates already exists on disk and that they are untouched.  I cannot be 100% sure that the configuration hasn't changed since Caddy saved the certificates some days ago, since I fiddled a lot with the configuration because of the situation that made me open this issue in the first place.

I guess I'll have to wait until I'm not rate-limited anymore and then try again with a clear slate.

**Comment by niv** [comment]
I think there might be a bug where it fails to write out the certificates it got when it requests a bunch and then runs into rate-limiting.

Unfortunately I can't give you a testcase since I've since ran into limits, but here is what I did:
- make it generate a toplevel cert for my domain
- restart server a few of times with different configs, testing things
- add about 10 vhosts, all subdomains under the same domain
- run into ratelimit at next restart, but no certs were requested

I'm sorry this isn't more helpful, but that's all I can offer as a user anecdote.

**Comment by captncraig** [comment]
I am also ratelimited. The final time I got a cert I had to subdomains added to my Caddyfile. One of them (`a.mydomain`) was issued a cert (it shows up in the transparency report) but that cert was never saved to disk because `b.subdomain` got rate limited. Fairly minor inconvenience, but I think caddy or lego could have found and save the cert for `a.mydomain` but didn't because there were errors.
